src/ucp/tag/offload.c

```diff
@@ -231,20 +233,143 @@ void ucp_tag_offload_cancel(ucp_worker_t *worker, ucp_request_t *req, int force)
+         }
+
+         status = ucp_request_memory_reg(context, UCS_BIT(mdi), req->recv.buffer,
+                                         length, req->recv.datatype,
+                                         &req->recv.state, req);
+         if (status != UCS_OK) {
+             return status;
+         }
+
+         req->recv.rdesc = NULL;
+         iov.buffer      = (void*)req->recv.buffer;
+         iov.memh        = req->recv.state.dt.contig.memh[0];
+     } else {
+         rdesc = ucp_worker_mpool_get(worker);
+         if (rdesc == NULL) {
+             return UCS_ERR_NO_MEMORY;
+         }
+
+         iov.memh        = ucp_memh2uct(rdesc->memh, mdi);
+         iov.buffer      = rdesc + 1;
+         req->recv.rdesc = rdesc;
+     }
+
+     iov.length = length;
+     iov.count  = 1;
+     iov.stride = 0;
+
+     req->recv.uct_ctx.tag_consumed_cb = ucp_tag_offload_tag_consumed;
+     req->recv.uct_ctx.completed_cb    = ucp_tag_offload_completed;
+     req->recv.uct_ctx.rndv_cb         = ucp_tag_offload_rndv_cb;
+
+     status = uct_iface_tag_recv_zcopy(wiface->iface, req->recv.tag.tag,
+                                       req->recv.tag.tag_mask, &iov, 1,
+                                       &req->recv.uct_ctx);
+     if (status != UCS_OK) {
+         /* No more matching entries in the transport. */
+         ucp_tag_offload_release_buf(req);
+         UCP_WORKER_STAT_TAG_OFFLOAD(worker, BLOCK_TAG_EXCEED);
+         return status;
+     }
+
+     UCP_WORKER_STAT_TAG_OFFLOAD(worker, POSTED);
+     req->flags          |= UCP_REQUEST_FLAG_OFFLOADED;
+     req->recv.tag.wiface = wiface;
+     ++wiface->post_count;
+     ucs_trace_req("recv request %p (%p) was posted to transport (rsc %d)",
+                   req, req + 1, wiface->rsc_index);
+     return UCS_OK;
+
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_tag_offload_post_sw_reqs(ucp_request_t *req, ucp_request_queue_t *req_queue)
+{
+    ucp_worker_t *worker = req->recv.worker;
+    ucs_queue_iter_t iter;
+    ucs_status_t status;
+    ucp_request_t *req_e;
+    ucp_worker_iface_t *wiface;
+    size_t max_post;
+
+    wiface = ucp_tag_offload_iface(worker, req->recv.tag.tag);
+    if (ucs_unlikely(wiface == NULL)) {
+        UCP_WORKER_STAT_TAG_OFFLOAD(worker, BLOCK_NO_IFACE);
+        return UCS_ERR_NO_RESOURCE;
+    }
+
+    max_post = wiface->attr.cap.tag.recv.max_outstanding - wiface->post_count;
+
+    /* If large enough buffer is being posted to the transport,
+     * try to post all unposted requests from the same TM queue before.
+     * Check that:
+     * 1. The receive buffer being posted is large enough (>= FORCE_THRESH)
+     * 2. There is no any request which can't be posted to the transport
+     *    (sender rank wildcard or non-contig type)
+     * 3. Transport tag list is big enough to fit all unposted requests plus
+     *    the one being posted */
+    if ((req->recv.length < worker->context->config.ext.tm_force_thresh) ||
```

Comment:
the first two fast checks could be done before checking the tag hash (e.g before calling ucp_tag_offload_iface)